### PR TITLE
chore(core): Add additional API checks 🎬

### LINKED ISSUE #9801: feat(core): Add API version number
The Keyman Core API usually doesn't change in breaking ways between Keyman versions, so it doesn't make sense to change the version number in the core library (_libkmnkbp0-0 / libkeymancore1_) with each Keyman release.

But then it would be good to document the Core API version somewhere. See [this](https://github.com/keymanapp/keyman/pull/9794#discussion_r1364983057) discussion.

A good indication when the Core API version number needs to be updated is the file `linux/debian/libkeymancore.symbols`. If that file changes the Core API version needs to be bumped:
- the minor version number needs to be bumped if new lines get added
- the major version number needs to be bumped if existing lines get changed or removed

The API verification step in `.github/workflows/deb-packaging.yml` could be enhanced to check that the API version got updated.

### PR DESCRIPTION
**Pull request by ermshiperete**
- Add additional checks for the symbols file:
  - if a line with a method name gets changed, the package version number in that line also needs to be updated
- Add additional checks for the API:
  - for major API changes (methods renamed or removed as reflected in the .symbols file) the API version
    numbers needs to be incremented
  - the API version numbers in the .symbols file and in `CORE_API_VERSION.md` need to match

Closes #9801

@keymanapp-test-bot skip

### CODE REVIEW

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Given the changes to trigger-builds impacts all platforms and every build, I would like to split them out into a separate branch for discussion and review. The changes shouldn't be buried.

I haven't yet reviewed the other aspects of this PR

**Comment by ermshiperete** [comment]
I split out the Core API version change into #9877.

File: linux/scripts/deb-packaging.sh [hunk 1]
```diff
@@ -11,14 +13,18 @@ THIS_SCRIPT="$(readlink -f "${BASH_SOURCE[0]}")"
...
-  "verify                    Verify API" \
-  "--gha                     Build from GitHub Action"
+  "dependencies               Install dependencies as found in debian/control" \
+  "source+                    Build source package" \
+  "verify                     Verify API" \
+  "--gha                      Build from GitHub Action" \
+  "--bin-pkg=BIN_PKG          Path and name of binary Debian package (for verify action)" \
+  "--pkg-version=PKG_VERSION  The version of the Debian package (for verify action)" \
+  "--git-ref=GIT_REF          The ref of the HEAD commit, e.g. HEAD of the PR branch (for verify action)" \
+  "--git-base=GIT_BASE        The ref of the base commit, e.g. HEAD of the master branch (for verify action)"
```
Comment (mcdurdin) [inline linux/scripts/deb-packaging.sh:23]: Should these be SHAs or commitishes/refs, tag names, branch names or what? :grin:

Comment (mcdurdin) [reply linux/scripts/deb-packaging.sh:23]:
```suggestion
  "--git-ref=GIT_REF          The SHA of the HEAD commit, e.g. of the PR branch (for verify action)" \
  "--git-base=GIT_BASE        The ref of the base commit, usually the name of the base branch, e.g. master, stable-16.0 (for verify action)"
```

File: linux/scripts/deb-packaging.sh [hunk 2]
```diff
@@ -50,19 +56,151 @@ source_action() {
...
+  # Checks that the API did not change compared to what's documented in the .symbols file
+  tmpDir=$(mktemp -d)
+  dpkg -x "${BIN_PKG}" "${tmpDir}"
+  cd debian
+  dpkg-gensymbols -v"${PKG_VERSION}" -p"${PKG_NAME}" -e"${tmpDir}"/usr/lib/x86_64-linux-gnu/"${LIB_NAME}".so* -O"${PKG_NAME}.symbols" -c4
+  output_ok "${LIB_NAME} API didn't change"
+  cd "${REPO_ROOT}/linux"
+}
+
+is_symbols_file_changed() {
```
Comment (mcdurdin) [inline linux/scripts/deb-packaging.sh:89]:
```suggestion
#
# Compare the SHA of the base and head commits for changes to the .symbols file
#
is_symbols_file_changed() {
```

File: linux/scripts/deb-packaging.sh [hunk 3]
```diff
@@ -50,19 +56,151 @@ source_action() {
...
+}
+
+output_error() {
+  echo ":x: $1" >&2
+  builder_echo error "ERROR: $1"
+}
+
+check_api_not_changed() {
+  # Checks that the API did not change compared to what's documented in the .symbols file
+  tmpDir=$(mktemp -d)
```
Comment (mcdurdin) [inline linux/scripts/deb-packaging.sh:81]: Cleanup?

File: linux/scripts/deb-packaging.sh [hunk 4]
```diff
@@ -50,19 +56,151 @@ source_action() {
...
+}
+
+check_updated_version_number() {
+  # Checks that the package version number got updated in the .symbols file if it got changed
+  # shellcheck disable=SC2310
+  if is_symbols_file_changed; then
+    # .symbols file changed, now check if the package version got updated as well
+    # Note: We don't check that ALL changes in that file have an updated package version -
+    # we hope this gets flagged in code review.
+    if ! git log -p -1 -- "debian/${PKG_NAME}.symbols" | grep -q "${PKG_VERSION}"; then
```
Comment (mcdurdin) [inline linux/scripts/deb-packaging.sh:106]: `${PKG_VERSION}` could just be `${VERSION}` I think? And then we can eliminate all of that from this script?

File: linux/scripts/deb-packaging.sh [hunk 5]
```diff
@@ -50,19 +56,151 @@ source_action() {
...
+  return 0
+}
+
+check_updated_version_number() {
+  # Checks that the package version number got updated in the .symbols file if it got changed
+  # shellcheck disable=SC2310
+  if is_symbols_file_changed; then
+    # .symbols file changed, now check if the package version got updated as well
+    # Note: We don't check that ALL changes in that file have an updated package version -
+    # we hope this gets flagged in code review.
```
Comment (mcdurdin) [inline linux/scripts/deb-packaging.sh:105]:
```suggestion
    # .symbols file changed, now check if the package version got updated as well
    # Note: We don't check that ALL changes in that file have an updated package version -
    # we hope this gets flagged in code review.
    # Note: This version number check may not match the actual released version, if the branch
    # is out of date when it is merged to the release branch (master/beta/stable-x.y). If this
    # is considered important, then make sure the branch is up to date, and wait for test
    # builds to complete, before merging.
```

File: linux/scripts/deb-packaging.sh [hunk 6]
```diff
@@ -50,19 +56,151 @@ source_action() {
...
+    fi
+  else
+    output_ok "${PKG_NAME}.symbols file didn't change"
+  fi
+}
+
+get_api_version_in_symbols_file() {
+  # Extract 1 from "libkeymancore.so.1 libkeymancore #MINVER#"
+  local firstline
+  firstline=$(head -1 "debian/${PKG_NAME}.symbols")
```
Comment (mcdurdin) [inline linux/scripts/deb-packaging.sh:120]:
```suggestion
  firstline="$(head -1 "debian/${PKG_NAME}.symbols")"
```

File: linux/scripts/deb-packaging.sh [hunk 7]
```diff
@@ -50,19 +56,162 @@ source_action() {
...
+  echo "${firstline}"
+}
+
+is_api_version_updated() {
+  local NEW_VERSION OLD_VERSION
+  git checkout "${GIT_SHA}" -- "debian/${PKG_NAME}.symbols"
+  NEW_VERSION=$(get_api_version_in_symbols_file)
+  git checkout "${GIT_BASE}" -- "debian/${PKG_NAME}.symbols"
+  OLD_VERSION=$(get_api_version_in_symbols_file)
+  git checkout "${GIT_SHA}" -- "debian/${PKG_NAME}.symbols"
```
Comment (mcdurdin) [inline linux/scripts/deb-packaging.sh:143]: We could reduce the number of checkouts by switching order of operations here, but it's not terribly important.

**Review by mcdurdin — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.210-alpha](https://keyman.com/downloads/releases/alpha/17.0.210)

**Pull request merged**